### PR TITLE
Make http_uri:encode/1 percent encoding conform to RFC 3986

### PR DESCRIPTION
**Pull request by richcarl**
Previously, only characters explicitly listed as reserved were being
escaped, while others outside the ASCII range, such as 'ö' were passed
through unchanged. Also, the reserved list was wrong, both missing some
characters (e.g. '!' and '*') and listing some that are not actually
reserved (e.g. '<' and '>').

The correct behaviour is to always escape all characters that are not in the
unreserved set, i.e., A-Z, a-z, 0-9, and '-', '_', '.', '~'.

### CODE REVIEW

File: lib/inets/src/http_lib/http_uri.erl [hunk 1]
```diff
@@ -302,19 +313,19 @@ path("") ->
...
     Path.
 
-uri_encode(Char, Reserved) ->
-    case sets:is_element(Char, Reserved) of
-	true ->
-	    [ $% | http_util:integer_to_hexlist(Char)];
+uri_encode(Char) ->
+    case is_unreserved(Char) of
 	false ->
+	    [ $% | http_util:integer_to_hexlist(Char)];
```
Comment (mikpe) [inline lib/inets/src/http_lib/http_uri.erl]: I see why you wrote it this way, but it leads to a double-negation (is_unreserved() = false) which makes it awkward to read.

Comment (richcarl) [reply lib/inets/src/http_lib/http_uri.erl]: I tried to minimize the diff, but it also matches the documentation: if not in the unreserved set, escape it, otherwise leave it as is.

Comment (richcarl) [reply lib/inets/src/http_lib/http_uri.erl]: Also, in the URI terminology, "unreserved" is a different thing from "not reserved".

**Comment by IngelaAndin** [comment]
If you want to improve the inets -httpd code I think it should be by starting to use uri_string instead of fixing this old code. We removed the use of http_uri in all other places in OTP.  Making  httpd  use uri_string internally (not removing http_uri) for backwards compatible reasons is the way forward.
I know @peterdmv  has some feedback for your other PR on uri_string that he will be posting soon.

**Comment by richcarl** [comment]
Fixed commit message: escape/1 -> encode/1.

**Comment by richcarl** [comment]
First, uri_string shows up in OTP 21, while this can be backported to OTP 20 and earlier. And second, it fixes an actual erroneous behaviour in a (still) documented module, and even though you plan on removing it eventually, I think that it's better if the final version of the module is actually correct.

**Comment by IngelaAndin** [comment]
How do you mean?  The uri_string module could be backported to OTP-20.  And if you change the http_uri module to use uri_string module internally it would work correctly.

**Comment by richcarl** [comment]
Would prefer a patch that does not add a whole module (which is undocumented in OTP 20).

**Comment by IngelaAndin** [comment]
PR are accepted for maint and master so we are interested in solutions that benefit those. How you possibly backport it to OTP-20 is not really our problem, if you do not want to create a new module you could always copy the implementation to an existing one.

**Comment by richcarl** [comment]
I don't really see the problem here. Do you want the last version of this module to still have this bug? A standalone fix like this is much better than inserting a dependency on a separate module.

**Comment by richcarl** [comment]
By the way, I see no obvious way to percent-encode a given URL by any function in the uri_string module. All the entry points seem to demand that you already have split it into a #uri_map, or that the URL is already properly percent-encoded.

**Comment by IngelaAndin** [comment]
Inets is already dependent on stdlib, so there is no new dependency. For your URI questions I see no reason to duplicate the discussion from  PR 2355

**Comment by richcarl** [comment]
Well, you suggest that I solve this problem by implementing the encode/1 function in terms of the functions in uri_string, but I see no apparent way of doing so - can you demonstrate how you think this should be done? (Also, PR 2355 is about unwanted decoding, while this PR is about necessary encoding.)

**Comment by peterdmv** [comment]
Let's take an example to see what this function is doing with this PR. The input is not a valid URI because the charater "ö" should be percent-encoded.
```erlang
1> http_uri:encode("http://localhos%74-örebro").
"http%3A%2F%2Flocalhos%2574-%F6rebro"
```
The result is something that looks strange enough to double-check it with our parser:
```erlang
2> uri_string:parse("http%3A%2F%2Flocalhos%2574-%F6rebro").
#{path => "http%3A%2F%2Flocalhos%2574-%F6rebro"}
```
So it is a valid URI with only a path component. It is not the same URI though as the input because even the percent sign has been encoded (%2574).

We can get a similar result when using the example on https://www.w3schools.com/tags/ref_urlencode.asp.

```
Your input was received as:
text=http%3A%2F%2Flocalhos%2574-%C3%B6rebro
```
On the browser line we see the URI for the result as
https://www.w3schools.com/action_page2.php?text=http%3A%2F%2Flocalhos%2574-%C3%B6rebro

The BNF of the query component is  
```
query       = *( pchar / "/" / "?" )
```

It allows slashes so the encoding is faulty in both cases. It is indeed not obvious how to implement this faulty behavior using the uri_string module. The question is how are you using this module? What are the input and output for your function?

**Comment by IngelaAndin** [comment]
Humm ... I have been digging into the history of this function. And I dare say that it was never based on RFC 3986. The reference to the RFC was added later to the module and so was the documentation of the function. Also later there where test added that assumes it works as follows "URL Encoding, a scheme in which text data transmitted to web servers is encoded by replacing spaces (" ") with plus signs ("+") and non alpha numeric characters are replaced by their ASCII equivalents." So I think there are plenty of room for misunderstandings in the life of this function (some of them might be mine, however I did not write the documentation). 
Most use of this function has been removed form OTP, and we now plan to remove all of its uses,
as we do not see that it works correctly. 

So Peters question is interesting how do you expect it work? Because the documented way seems to me, not to be valid.

**Comment by IngelaAndin** [comment]
Let me put it this way.  To accept a PR for http_uri:encode we want the PR to clearly improve the function. It is unfortunate that this function was documented the way it was, and probably it was
due to lack of good routines back in the days. But there is not point in pointing fingers we want to
move on in a good direction. That is the documentation will have to be changed to describe a possible sensible use of it under circumstances that makes sense. We will also like some test at minimum for your failing use case.  However we are trying to move away from a lot of legacy code in inets and uri_string is the favored API handle URIs.

**Comment by IngelaAndin** [comment]
I am closing this PR for lack of activity, if you are interested in changing the PR as described you are welcome to reopen.
